Posts / State: Add post counts to global state

aduth opened the pull request:
This pull request seeks to implement post counts behaviors in the global Redux application state. This is planned to be used as part of the focus on Custom Post Types, specifically in displaying counts for each status for a post type in the list view filter bar.

__Testing instructions:__

The post counts state is not used yet, but you should verify that the inclusion of the new `state.posts.counts` subtree does not introduce regressions to the application.

Ensure that Mocha tests pass by running `npm run test-client`.

__Open questions:__

- The state is implemented using WP-API undocumented post types endpoint. @timmyc @nylen : Do you agree that this is the correct approach to be using here, as you've started to roll out in the editor (#4107)?

nylen commented:
Yes, I think this is a good use of the new API infrastructure.

Code looks good to my (admittedly non-Redux-expert) eyes.  :+1: 

Worth noting that we're in the process of making a change to the URL structures of these endpoints - waiting on other client code, though, so not ready to do it here:

```diff
- /wp-json/sites/2916284/wpcom/v2/post-counts/post
+ /wpcom/v2/sites/2916284/post-counts/post
```

aduth commented:
@nylen : Cool, should only require changes to the undocumented endpoint here then:

https://github.com/Automattic/wp-calypso/blob/a3c2e36/client/lib/wpcom-undocumented/lib/site.js#L186-L196

(Originally added by @timmyc in #3836)

aduth commented:
Rebased and squashed, will plan to merge after tests pass and another quick test of behavior locally.

timmyc commented on client/state/posts/counts/actions.js:
```diff
@@ -0,0 +1,64 @@
+ * @param  {String}   postType Post type
+ * @return {Function}          Action thunk
+ */
+export function requestPostCounts( siteId, postType ) {
```
Though not used in calypso, just noting that the endpoint does support passing in an `author(int)` to query the API for post-counts for a specific author [#](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/post-counts/%24post_type/)

aduth replied:
>Though not used in calypso, just noting that the endpoint does support passing in an `author(int)` to query the API for post-counts for a specific author [#](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/post-counts/%24post_type/)

I might see how this could be useful for the case of filtering "my" vs. "everyone" posts, though not sure there's a huge performance win for querying only the author type, and would complicate the implementation quite a bit more.

Should a need arise in the future, it wouldn't be a stretch to imagine adding an additional argument or separate action creator for querying post counts specific to one author.

timmyc replied:
> would complicate the implementation quite a bit more

Totally agree, just pointing it out for reference - and since the endpoint is fresh in my mind from the WP-API work.

aduth commented:
>Rebased and squashed, will plan to merge after tests pass and another quick test of behavior locally.

Whoops, must have got distracted :smile: 

I rebased once more, found that tests had broken due to merging of route changes in #4401, and rebased again to update the paths to the new structure.
